src/book-test-instance.py

```diff
@@ -0,0 +1,393 @@
+    """
+
+    if os.getenv('JIRA_APP_KEY'):
+        return {
+            'type': 'oauth',
+            'auth': OAuth1(
+                os.getenv('JIRA_APP_KEY'),
+                os.getenv('JIRA_APP_SECRET'),
+                os.getenv('JIRA_OAUTH_TOKEN'),
+                os.getenv('JIRA_OAUTH_TOKEN_SECRET')
+            )
+        }
+
+    return {
+        'type': 'basic',
+        'auth': HTTPBasicAuth(os.getenv('JIRA_USER'), os.getenv('JIRA_PASS'))
+    } if os.getenv('JIRA_USER') else {}
+
+
+def save_results(results, filename='booking-results.json'):
+    with open(filename, 'w') as results_file:
+        json.dump(results, results_file, indent=2)
+
+
+"""
+
+Main
+
+"""
+
+if __name__ == '__main__':
+
+    # Options
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--pr-url",
+                        help="pull request URL")
+    parser.add_argument("-n", "--dryrun", action="store_true",
+                        help="gives a course of action but doesn't execute")
+    parser.add_argument("-v", "--verbosity", action="count", default=0,
+                        help="increase output verbosity")
+    parser.add_argument("--no-cache", action="store_true",
+                        help="Disable request cache use")
+    parser.add_argument("--no-booking", action="store_true",
+                        help="Disable instance booking action")
+    parser.add_argument("--results", default="booking-results.json",
+                        help="Save result in json to the specified file (default booking-results.json)")
+    args = parser.parse_args()
+
+    # Parsed options
+    pr_url = args.pr_url if args.pr_url else os.getenv('CIRCLE_PULL_REQUEST')
+    dryrun = args.dryrun
+    verbose = args.verbosity
+    use_request_cache = not args.no_cache
+    results_file = args.results
+    vprint = print if verbose else lambda *a, **k: None
+    # Auth
+    jira_auth = get_jira_auth()
+
+    # Main program
+
+    vprint('\n# Running for {0}'.format(pr_url))
+    if dryrun:
+        vprint('## Dry run, nothing will be commited.')
+
+    # Fetch PR details
+    pull = get_pull_request(pr_url=pr_url)
+    if not pull:
+        raise Exception('No pull request found, aborting.')
+
+    # Fetch issue details from Github PR
+    try:
+        issue = get_jira_issue(pr=pull)
+    except Exception as e:
+        vprint(e)
+        issue = False
+
+    # Define instance
+    if not issue:
+        instance = False
+        vprint('No corresponding issue found, booking will not be executed.')
+    else:
+        if issue['test_instance']:
+            instance = issue['test_instance']
+            vprint(
+                'Issue is already deployed on {0}, reusing.'.format(instance))
+        else:
+            instance = get_available_instance()
+
+        # Book instance
+        try:
+            book_instance(instance, issue)
+        except Exception as e:
+            vprint(e)
+            instance = False
+
+    save_results({
```

Comment:
For now the CI actually only needs the instance name, so it would be simpler in usage if the script returns this as a string on success, instead of writing a file which needs to be parsed only to get a single string out of it.